src/Ssl/CSRGenerator.php

```diff
@@ -0,0 +1,108 @@
+<?php
+
+/*
+ * This file is part of the ACME PHP library.
+ *
+ * (c) Titouan Galopin <galopintitouan@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace AcmePhp\Core\Ssl;
+
+use AcmePhp\Core\Ssl\Exception\GeneratingCsrFailedException;
+use Psr\Log\LoggerInterface;
+use Psr\Log\NullLogger;
+
+/**
+ * Generate CSR.
+ *
+ * @author Jérémy Derussé <jeremy@derusse.com>
+ */
+class CSRGenerator
+{
+    /** @var LoggerInterface */
+    private $logger;
+
+    /**
+     * @param LoggerInterface|null $logger
+     */
+    public function __construct(LoggerInterface $logger = null)
+    {
+        $this->logger = $logger ?: new NullLogger();
+    }
+
+    /**
+     * Generate CSR string from given parameters.
+     *
+     * @param string  $domain
+     * @param KeyPair $domainKeyPair
+     * @param CSR     $csr
+     *
+     * @return string
+     */
+    public function generateCSR($domain, KeyPair $domainKeyPair, CSR $csr)
+    {
+        $csrData = $csr->toArray();
+        $csrData['commonName'] = $domain;
+        $privateKey = $domainKeyPair->getPrivateKey();
+
+        $this->logger->debug('Generating Certificate Signing Request ...', ['csrData' => $csrData]);
+
+        $subjectAltName = $csr->getSubjectAlternativeNames();
+        if (!in_array($domain, $subjectAltName)) {
+            $subjectAltName[] = $domain;
+        }
+
+        if (1 < count($subjectAltName)) {
+            $sslConfigTemplate = <<<'EOL'
+[ req ]
+distinguished_name = req_distinguished_name
+req_extensions = v3_req
+[ req_distinguished_name ]
+[ v3_req ]
+basicConstraints = CA:FALSE
+keyUsage = nonRepudiation, digitalSignature, keyEncipherment
+subjectAltName = @req_subject_alt_name
+[ req_subject_alt_name ]
+%s
+EOL;
+            $sslConfigDomains = [];
+            foreach (array_values($subjectAltName) as $index => $domain) {
```

Comment:
I usually put spaces before and after foreach (same for try). I'm not sure it's in any convention, but I would like to keep code style consistent :) .